protocol/contracts/nexus/Shard.sol

```diff
@@ -0,0 +1,555 @@
+    address private constant ZERO_ADDRESS = 0x0000000000000000000000000000000000000000;
+
+    /// @notice all 3 mappings below not combined into a struct for reasons allowedPartnerShardIds read often than rest. 
+    /// also for easy public reads(read functions). avoids cycle error
+    /// choosing mapping of address to EnumerableSet to help supporting view functions like getPartnerMintInfo
+    mapping(address => EnumerableSet.UintSet) private allowedPartnerShardIds;
+    mapping(address => mapping(uint256 => uint256)) public allowedPartnersShardCaps;
+    mapping(address => mapping(uint256 => uint256)) public partnerMintBalances;
+    /// @notice keep track of all shards reserved for partners. helps to guard general mints also
+    EnumerableSet.UintSet private partnerShards;
+    /// @notice keep track of all other shards not reserved for partners
+    EnumerableSet.UintSet private regularShards;
+
+    /// @notice shard ids to uris
+    mapping(uint256 => string) public shardUris;
+
+    /// @notice minters who can mint for templars
+    mapping(address => bool) public templarMinters;
+
+    mapping(uint256 => Recipe) private recipes;
+
+    /// @notice each shard belongs to exactly 1 enclave. an enclave can have many shards
+    mapping(uint8 => EnumerableSet.UintSet) private enclaveShards;
+
+    /// @notice Enclave types
+    enum Enclave {
+        Chaos,
+        Mystery,
+        Logic,
+        Order,
+        Structure
+    }
+
+    /// @notice for compact view functions
+    struct PartnerMintInfo {
+        uint256[] shardIds;
+        uint256[] caps;
+        uint256[] balances;
+    }
+
+    /// @dev maintaining uint256[] to avoid casting in batch functions like burnBatch
+    struct Recipe {
+        uint256[] inputShardIds;
+        uint256[] inputShardAmounts;
+        uint256[] outputShardIds;
+        uint256[] outputShardAmounts;
+    }
+
+    event Transmuted(address caller, uint256 recipeId);
+    event ShardMinted();
+    event ShardUriSet(uint256 shardId, string uri);
+    event RecipeSet(uint256 recipeId, Recipe recipe);
+    event PartnerAllowedShardIdSet(address partner, uint256 shardId, bool allow);
+    event PartnerAllowedShardIdsSet(address partner, uint256[] shardIds, bool[] allowances);
+    event PartnerAllowedShardCapsSet(address partner, uint256[] shardIds, uint256[] caps);
+    event TemplarMinterSet(address minter, bool allowed);
+    event RecipeDeleted(uint256 recipeId);
+    event ShardEnclaveSet(Enclave enclave, uint256 shardId);
+    event RegularShardIdSet(uint256 shardId, bool allow);
+
+    error ReservedPartnerShard(uint256 shardId);
+    error ShardMintNotAllowed(address caller, uint256 shardId);
+    error CannotTransmute(address caller);
+    error CannotMint(address caller);
+    error MintCapExceeded(uint256 cap, uint256 amount);
+    error InvalidRecipe(uint256 recipeId);
+    error ZeroAddress();
+    error InvalidParamLength();
+    error InvalidAccess(address caller);
```

Comment:
Perhaps some dup of errors with `CommonEventsAndErrors`?

Reply:
`CommonEventsAndErrrors.InvalidAccess()` takes no arguments